ih-core/bin/ih-setup

```diff
@@ -0,0 +1,334 @@
+        DAG="$DAG
+$DEP $STEP"
+      done
+    fi
+  done
+
+  SORTED=$(echo "$DAG" | tsort | tr '\n' ' ')
+
+  echo "$SORTED"
+}
+
+function ih::private::load-steps() {
+
+  local STEP_FILES
+  local STEP_FILE
+
+  STEP_FILES=$(grep -rlE 'function ih::setup::[A-z0-9_-]+::install' "$IH_CORE_LIB_DIR")
+
+  for STEP_FILE in $STEP_FILES; do
+    ih::log::debug "Steps found in $STEP_FILE"
+    # shellcheck disable=SC1090
+    . "$STEP_FILE"
+  done
+}
+
+# Applies command $1 to the steps listed in the rest of the arguments.
+# If there are no steps listed then the command will be applied to
+# all the known steps.
+function ih::private::apply-to-steps() {
+  local COMMAND=$1
+  shift
+  STEPS=$*
+  if [[ -z $STEPS ]]; then
+    STEPS=$(ih::private::list-steps)
+  fi
+  for STEP in $STEPS; do
+    "ih::private::$COMMAND" "$STEP"
+  done
+}
+
+# Prints the description of the step in $1
+# along the the deps.
+function ih::private::describe-step() {
+  blue "--------- $1 -----------\n"
+  ih::private::step-help "$1"
+  echo ""
+  echo "depends on: $(ih::private::step-deps "$1")"
+  echo ""
+}
+
+# Tests the step $1 with nicer logging.
+# If IH_DEBUG is not set to 1 then this will
+# discard the output of the test command for the step.
+function ih::private::test-step() {
+  ih::log::debug "Testing step $STEP..."
+  if [[ $IH_DEBUG = "1" ]]; then
+    ih::private::step-test "$STEP"
+  else
+    ih::private::step-test "$STEP" >/dev/null
+  fi
+  TEST_CODE=$?
+  if [[ $TEST_CODE = 0 ]]; then
+    green "$STEP has been installed\n"
+  else
+    red "$STEP needs to be installed\n"
+  fi
+}
+
+# Tests the dependencies of the step in $1,
+# recursively. Returns 1 if any dependency
+# is not installed.
+function ih::private::test-deps() {
+  local DEPTH="$2  "
+  local FAILED=0
+  local DEP
+  local DEPS
+  DEPS=$(ih::private::try-invoke "$1" deps)
+  for DEP in $DEPS; do
+    ih::private::step-test "$DEP" >/dev/null
+    TEST_CODE=$?
+    if [[ $TEST_CODE != 0 ]]; then
+      red "$DEPTH Dependency $DEP not installed\n"
+      FAILED=1
+    else
+      green "$DEPTH Dependency $DEP is installed\n"
+    fi
+    ih::private::test-deps "$DEP" "$DEPTH"
+    TEST_CODE=$?
+    if [[ $TEST_CODE != 0 ]]; then
+      FAILED=1
+    fi
+  done
+
+  if [[ $FAILED = "1" ]]; then
+    return 1
+  fi
+  return 0
+
+}
+
+# Calls install on the steps passed as arguments.
+# If IH_FORCE is set to 1 it will install each step; otheriwise it will skip steps where
+# assert-installed returns 0
+# If IH_NO_CONFIRM is set to 1 it will skip confirmation; otherwise it will ask the user to confirm
+# before invoking the step.
+function ih::private::install-step() {
+  local TEST_CODE
+  if [ "$IH_FORCE" != "1" ]; then
+    ih::log::debug "Checking whether step $STEP needs to be installed..."
+    ih::private::step-test "$STEP" >/dev/null
+    TEST_CODE=$?
+    if [[ $TEST_CODE -eq 0 ]]; then
+      ih::log::info "${STEP} has been installed (use --force to re-install)"
```

Comment:
AFAICT the arg parser only accepts `-f`, not `--force`, right?

Reply:
Right